### PR TITLE
Softfloat 3a

### PR DESCRIPTION
**Pull request by jenswi-linaro**
Importing and integrating Softfloat 3a to provide floating point support for user TAs

### CODE REVIEW

**Comment by pascal-brand-st-dev** [comment]
Overall looks good. Some general questions:

* should we compile softfload in all cases? Or would we rather have a new `CFG_FLOAT_SUPPORT=y`
* is the softfload code arm-specific?
* softfloat looks compiled only on arm32? will the new os_test be ok with 64bits TA? or should we only test when `CFG_FLOAT_SUPPORT=y` (if we go for it).

**Comment by jenswi-linaro** [comment]
The softfloat code itself isn't arm or arm32 specific, but it's only used for arm32 (will never be used for arm64 as there's not soft float ABI defined) so I didn't bother putting it in some common directory.

With this PR we'll have floating point support for 32-bit TAs, but not for 64-bit TAs. I think we should have floating point support for all TAs, what we'll configure is whether is should be soft-float or hard-float. Since we don't have hard-float yet, I guess there's not a need to configure kind of float.

I'm starting to work on getting hard float support for 32-bit and 64-bit TAs. In the process I'll probably need to do something about the platform specific `platform_flags.mk`.

**Comment by jforissier** [comment]
On Fri, Dec 11, 2015 at 10:14 AM, Jens Wiklander <notifications@github.com>
wrote:

> The softfloat code itself isn't arm or arm32 specific, but it's only used
> for arm32 (will never be used for arm64 as there's not soft float ABI
> defined) so I didn't bother putting it in some common directory.
>
> With this PR we'll have floating point support for 32-bit TAs, but not for
> 64-bit TAs. I think we should have floating point support for all TAs, what
> we'll configure is whether is should be soft-float or hard-float. Since we
> don't have hard-float yet, I guess there's not a need to configure kind of
> float.
>
+1 but what if someone just wants to remove the fp support this patch is
adding, for whatever reason? (maybe to reduce binary size or to use a
custom float library). Would that make sense ?
If so, I wouldn't mind having e.g., CFG_LIBUTILS_ARM32_SOFTFLOAT_SUPPORT ?=
y to enable this patch (this is a long name but I think it's important to
have descriptive names for all the non-trivial stuff, or users may get
lost).

> I'm starting to work on getting hard float support for 32-bit and 64-bit
> TAs. In the process I'll probably need to do something about the platform
> specific platform_flags.mk.
>
+1
If you can find some way of moving what is actually *not* platform-specific
in these files to a common place, that would be great :)

**Comment by jenswi-linaro** [comment]
I can see why one would like to disable all floating point support just to make sure that nothing unexpected sneaks into a TA. The soft-float code isn't any overhead if not used as it will not be pulled into the TA then, so it's not about sizes of the TA.

It sounds a bit hypothetical with replacing the soft-float implementation, I'd like to put that in the `CFG_LIBUTILS_WITH_ISOC` basket. It could very well go hand in hand.

If you agree I'll add a CFG_FLOAT_SUPPORT=y that will enable floating point in a TA. To guard against accidental usage of floating point in a TA simply unset this.

**Comment by jforissier** [comment]
On Fri, Dec 11, 2015 at 10:51 AM, Jens Wiklander <notifications@github.com>
wrote:

> I can see why one would like to disable all floating point support just to
> make sure that nothing unexpected sneaks into a TA. The soft-float code
> isn't any overhead if not used as it will not be pulled into the TA then,
> so it's not about sizes of the TA.
>
True.

> It sounds a bit hypothetical with replacing the soft-float implementation,
> I'd like to put that in the CFG_LIBUTILS_WITH_ISOC basket. It could very
> well go hand in hand.
>
+1

> If you agree I'll add a CFG_FLOAT_SUPPORT=y that will enable floating
> point in a TA. To guard against accidental usage of floating point in a TA
> simply unset this.
>
For now this is good indeed (add "TA" in the name maybe?).
I'm just a bit worried how this will evolve when hard float support is
added for 32- and 64-bit TAs. The name will get ambiguous and may have to
change, or be complemented with other variables, which is probably not a
big issue anyway. So feel free to keep this name.

**Comment by jenswi-linaro** [comment]
Added `CFG_TA_FLOAT_SUPPORT` flag

**Comment by jenswi-linaro** [comment]
Propagate `CFG_TA_FLOAT_SUPPORT` to TA dev kit.

**Comment by jenswi-linaro** [comment]
Sorry about the mess, propagating CFG_TA_FLOAT_SUPPORT to TA dev kit has to wait until I've fixed hard-float.

**Comment by jenswi-linaro** [comment]
Rebased.
Enables CFG_TA_FLOAT_SUPPORT for arm32 platforms.
CFG_TA_FLOAT_SUPPORT is also exported to TA dev kit to be able to select if floating point should be tested or not.

File: lib/libutils/isoc/arch/arm/arm32_aeabi_shift.c [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/*
```
Comment (pascal-brand-st-dev) [inline lib/libutils/isoc/arch/arm/arm32_aeabi_shift.c]: Is it the correct license?

**Comment by pascal-brand-st-dev** [comment]
Apart from the license question,

    Reviewed-by: Pascal Brand <pascal.brand@linaro.org>

File: lib/libutils/isoc/arch/arm/arm32_aeabi_shift.c (hunk 1 above)
Comment (jenswi-linaro) [inline lib/libutils/isoc/arch/arm/arm32_aeabi_shift.c]: Yes, I wrote the code.

**Comment by jenswi-linaro** [comment]
Applied tags

**Pull request merged**